added SitemapLoader

aaishikdutta opened the pull request:
Relates to https://github.com/embedchain/embedchain/issues/46. Added a SitemapLoader class that iterates over a list of urls from an XML sitemap and internally calls the WebPageLoader to load the urls. It is paired with the WebPageChunker to produce embedding for an entire website in one go.

gasolin commented:
@aaishikdutta  looks cool, try to run it but got error:

`AttributeError: 'SitemapLoader' object has no attribute 'create_chunks'`

might missing the `embedchain/chunkers/site_map.py`

aaishikdutta commented:
@gasolin can you please paste your code so that I can reproduce. Thanks

gasolin commented:
@aaishikdutta sorry false alert, I copied chunkers to SitemapLoader instead of WebPageChunker 😅

aaishikdutta commented:
all good @gasolin. Let me know if you find any other issues in this.

cachho commented:
Can you write an exception for paths that point to images? Some sitemaps contain them.

aaishikdutta commented:
Hey @cachho are you referring to this syntax. Sorry, haven't worked with these sitemaps previously:

`<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
        xmlns:image="http://www.google.com/schemas/sitemap-image/1.1">
  <url>
    <loc>https://example.com/sample1.html</loc>
    <image:image>
      <image:loc>https://example.com/image.jpg</image:loc>
    </image:image>
    <image:image>
      <image:loc>https://example.com/photo.jpg</image:loc>
    </image:image>
  </url>
  <url>
    <loc>https://example.com/sample2.html</loc>
    <image:image>
      <image:loc>https://example.com/picture.jpg</image:loc>
    </image:image>
  </url>
</urlset>`

cachho commented:
> Hey @cachho are you referring to this syntax. Sorry, haven't worked with these sitemaps previously:
> 
> ![Screenshot_2023-07-07-08-31-06-41_40deb401b9ffe8e1df2f1cc5ba480b12.jpg](https://user-images.githubusercontent.com/107566376/251624399-aa3c83a4-fad9-4ee9-a342-a1fc4f529bee.jpg)

I haven't seen that one, I guess you're just taking the `<loc>` so that images would be excluded anyways, right?

No I'm talking about stuff like this: https://en.wikipedia.org/wiki/File:Streathamladyjayne.jpg
Where File is in the URL, because they can't be handled by the embedding model anyways, as far as I know.

But I see how that would be too individual. I might have to filter the sitemap myself.

cachho commented:
forget what I said, this is fine.

deshraj reviewed (commented):
Looks good. Can you please resolve the merge conflicts and run `make format` to the code based on repository guidelines? Thanks!

cachho commented:
resolved the merge conflicts for you

cachho reviewed (changes requested):
Please add config and run `make format lint`

aaishikdutta commented:
> Looks good. Can you please resolve the merge conflicts and run `make format` to the code based on repository guidelines? Thanks!

updated @deshraj 


> Please add config and run `make format lint`

updated @cachho

deshraj commented on embedchain/loaders/site_map.py:
```diff
@@ -0,0 +1,33 @@
+            return output
+
+        else:
+            raise response.raise_for_status()
```
You can make the code a lot simpler by doing something like this:

```python
import requests
from bs4 import BeautifulSoup
from embedchain.loaders.web_page import WebPageLoader


class SitemapLoader:
    def load_data(self, sitemap_url):
        """
        This method takes a sitemap URL as input and retrieves
        all the URLs to use the WebPageLoader to load content
        of each page.
        """
        output = []
        web_page_loader = WebPageLoader()
        response = requests.get(sitemap_url)
        response.raise_for_status()

        soup = BeautifulSoup(response.text, "xml")
        links = [link.text for link in soup.find_all("loc")]
        for link in links:
            each_load_data = web_page_loader.load_data(link)
            output.append(each_load_data)
        return [data[0] for data in output]
```

Make sure to run `make format` after making the change.

aaishikdutta replied:
updated @deshraj

deshraj commented on README.md:
````diff
@@ -294,6 +294,13 @@ To supply your own QnA pair, use the data_type as `qna_pair` and enter a tuple.
 ```
+### Sitemap
+
+To add a XML site map containing list of all urls, use the data_type as `site_map` and enter the sitemap url. Eg:
````
```suggestion
To add a XML site map containing list of all urls, use the data_type as `sitemap` and enter the sitemap url. Eg:
```

deshraj commented on README.md:
````diff
@@ -294,6 +294,13 @@ To supply your own QnA pair, use the data_type as `qna_pair` and enter a tuple.
+To add a XML site map containing list of all urls, use the data_type as `site_map` and enter the sitemap url. Eg:
+
+```python
+app.add('site_map', 'a_valid_sitemap_url/sitemap.xml')
````
```suggestion
app.add('sitemap', 'a_valid_sitemap_url/sitemap.xml')
```

deshraj commented on embedchain/data_formatter/data_formatter.py:
```diff
@@ -39,6 +40,7 @@ def _get_loader(self, data_type, config):
             "qna_pair": LocalQnaPairLoader(),
             "text": LocalTextLoader(),
             "docx": DocxFileLoader(),
+            "site_map": SitemapLoader(),
```
```suggestion
            "sitemap": SitemapLoader(),
```

deshraj commented on embedchain/data_formatter/data_formatter.py:
```diff
@@ -60,6 +62,7 @@ def _get_chunker(self, data_type, config):
             "qna_pair": QnaPairChunker(config),
             "text": TextChunker(config),
             "docx": DocxFileChunker(config),
+            "site_map": WebPageChunker(config),
```
```suggestion
            "sitemap": WebPageChunker(config),
```

deshraj reviewed (approved):
Great, thanks for adding.
